lib/rspec/matchers/built_in/has.rb

```diff
@@ -80,11 +71,7 @@ def predicate_matches?(value=true)
         end
 
         def root
-          # On 1.9, there appears to be a bug where String#match can return `false`
-          # rather than the match data object. Changing to Regex#match appears to
-          # work around this bug. For an example of this bug, see:
-          # https://travis-ci.org/rspec/rspec-expectations/jobs/27549635
-          self.class::REGEX.match(@method_name.to_s).captures.first
+          self.class::REGEX.match(@method_name).captures.first
```

Comment:
This comment implies there was a better historical implementation, presumedly using a string rather than a regex, if you're going to change the implementation in a non compatible way might as well revert to that?

Reply:
I didn't see a good reason for this change:
```diff
- self.class::REGEX.match(@method_name).captures.first
+ @method_name.match(self.class::REGEX).captures.first
```

The reason for dropping `to_s` is that Regex#match works fine with a Symbol argument, but apparently it didn't before 2.0.
Also, Symbol#match works, so we actually can use (revert to some original) `@method_name.match` if we decide to.

I don't have any preference here. Just took the option with less churn.

Reply:
Can we benchmark the various options and see which is best?

Reply:
```ruby
require 'benchmark/ips'

BE_PREDICATE_REGEX = /^(?:be_(?:an?_)?)(.*)/
HAS_REGEX = /^(?:have_)(.*)/

BE_METHOD_NAMES = [:be_an_apple, :be_apple]
HAS_METHOD_NAMES = [:have_apples, :have_apple]

# Sanity checks
BE_METHOD_NAMES.each do |method_name|
  fail unless BE_PREDICATE_REGEX.match(method_name).captures == method_name.match(BE_PREDICATE_REGEX).captures
end

HAS_METHOD_NAMES.each do |method_name|
  fail unless HAS_REGEX.match(method_name).captures == method_name.match(HAS_REGEX).captures
end

report = Benchmark.ips do |x|
  x.report('Regex#match') do
    BE_METHOD_NAMES.each do |method_name|
      BE_PREDICATE_REGEX.match(method_name).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      HAS_REGEX.match(method_name).captures.first
    end
  end

  x.report('Regex#match(to_s)') do
    BE_METHOD_NAMES.each do |method_name|
      BE_PREDICATE_REGEX.match(method_name.to_s).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      HAS_REGEX.match(method_name.to_s).captures.first
    end
  end

  x.report('Symbol#match') do
    BE_METHOD_NAMES.each do |method_name|
      method_name.match(BE_PREDICATE_REGEX).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      method_name.match(HAS_REGEX).captures.first
    end
  end
end

puts report
```

```
Calculating -------------------------------------
         Regex#match    154.233k (± 7.0%) i/s -    768.576k in   5.010355s
   Regex#match(to_s)    131.795k (± 7.5%) i/s -    657.359k in   5.019755s
        Symbol#match    148.697k (± 7.6%) i/s -    743.634k in   5.034812s
```
Consistent across runs, `Regex#match` is the winner.